### ISSUE TITLE
PEP 647: overloads for filter() don't work

### ISSUE THREAD

**Issue by gvanrossum** [microsoft/pyright#1356]
**Describe the bug**
A type variable (in the example, 'R') seems to spill out of an overload involving a type guard.

**To Reproduce**
```py
from typing import overload, Any, Callable, Iterable, Iterator, List, Optional, TypeVar
from typing_extensions import TypeGuard

T = TypeVar("T")
R = TypeVar("R")

@overload
def filter(f: Callable[[T], TypeGuard[R]], it: Iterable[T]) -> Iterator[R]: ...
@overload
def filter(f: Callable[[T], object], it: Iterable[T]) -> Iterator[T]: ...
def filter(f, it):
    for x in it:
        if f(x):
            yield x

def is_int(a: object) -> bool:
    return isinstance(a, int)

def main(a: List[object]) -> None:
    aa = filter(is_int, a)
    reveal_type(aa)
```

**Expected behavior**
Revealed type on last line should be `Iterator[object]`.

**Actual output**
```
  21:17 - info: Type of "aa" is "Iterator[R@filter]"
```

**VS Code extension or command-line**
This is directly on the command line, with no flags.

**Comment by erictraut** [comment]
I think you meant for the return type of `is_int` to be `TypeGuard[int]`. If I make that change, it works as expected.

**Comment by erictraut** [comment]
Oh, I see what you mean. Yeah, I agree that's a bug.

**Comment by erictraut** [comment]
This will be fixed in the next version of Pyright & Pylance. The fix also addresses the issue you noted in the typing sig thread where a subclass override of a method that returns TypeGuard was not flagged as incompatible.

**Comment by erictraut** [comment]
This is now fixed in Pyright 1.1.101, which I just published. It will also be included in the next release of Pylance.

**Comment by gvanrossum** [comment]
Thanks!
